### PR TITLE
[202012][Arista] Fix cmdline generation during warm-reboot from 201811/201911

### PR DESCRIPTION
**Pull request by Staphylo**
#### Why I did it

When performing a `warm-reboot` or `fast-reboot` from `201811` or `201911` to `202012` the kernel command line contains duplicate information.
This issue is related to a change that was made to make `202012` `boot0` file more futureproof.
A cold reboot brings everything back into a clean slate though not always desirable.

#### How I did it

Added some logic to properly detect the end of the Aboot cmdline when `cmdline-aboot-end` delimiter is not set (clean case)
Added some logic to regenerate the Aboot cmdline when `cmdline-aboot-end` is set but duplicate parameters exists before (dirty case)
Reorganized  some code to handle duplicate parameter handling in the allowlist.

#### How to verify it

Clean case (device only ran and installed 201811 and 201911, first installation of 202012)
 - Cold boot into `201811`
 - Install `202012`
 - Make sure `/host/kernel-params-base` does not contain image specific parameters.

Dirty case (device was already migrated from 201811/201911 to 202012):
 - Cold boot into `201811`
 - Install `202012` (without the fix)
 - `warm-reboot` into `202012`
 - Verify that `/proc/cmdline` contains duplicate parameters and `/host/kernel-params-base` contains image specific parameters
 - Install `202012` (with the fix)
 - Verify that `/host/kernel-params-base` is repaired
 - `warm-reboot` into new `202012`
 - Verify that `/proc/cmdline` no longer has duplicate parameters

#### Description for the changelog
Fix cmdline generation during warm-reboot from 201811/201911 to 202012

### CODE REVIEW

File: files/Aboot/boot0.j2
Comment (vaibhavhd) [inline files/Aboot/boot0.j2:89]: Why is this PR introducing a new option `hwaddr_ma1`?

File: files/Aboot/boot0.j2
Comment (vaibhavhd) [inline files/Aboot/boot0.j2:399]: What is this change intended for? Is it relevant to this PR's context of fixing duplicate options?

File: files/Aboot/boot0.j2
Comment (vaibhavhd) [inline files/Aboot/boot0.j2:663]: Is it always going to be safe to depend on `log_buf_len=.* $delimiter` to detect append bug? I mean hardcoding this check to look for `log_buf_len` can possibly lead to false result?

File: files/Aboot/boot0.j2
Comment (vaibhavhd) [inline files/Aboot/boot0.j2:674]: Is this comment correct? I think in this case, cmdline will have delimiter, but along with append bug. Please confirm.

File: files/Aboot/boot0.j2
Comment (vaibhavhd) [inline files/Aboot/boot0.j2:676]: This would remove some unique params from the cmdline too. I am assuming we only want to remove duplicates. Is this intentional?

```
$ cat temp_cmdline
reboot=p console=ttyS0 acpi=on Aboot=Aboot-norcal7-7.2.0-pcie2x4-6128821 block_flash=pci0000:00/0000:00:1f.2/.*host./target0:0:0/.*$ block_drive=pci0000:00/0000:00:1f.2/.*host./target2:0:0/.*$ net_ma2=pci0000:00/0000:00:01.0/.*$ net_ma1=pci0000:00/0000:00:01.1/.*$ block_usb2=pci0000:00/0000:00:14.0/\(usb3/3-2\|usb4/4-2\)/.*$ block_usb1=pci0000:00/0000:00:14.0/\(usb3/3-3\|usb4/4-5\)/.*$ platform=rook scd.lpc_irq=7 scd.lpc_res_addr=0xb0000000 scd.lpc_res_size=0x10000 sid=Gardena iommu=on intel_iommu=on tsc=reliable pcie_ports=native rhash_entries=1 usb-storage.delay_use=0 reassign_prefmem varlog_size=4096 net.ifnames=0 log_buf_len=1M iommu=on intel_iommu=on tsc=reliable pcie_ports=native rhash_entries=1 usb-storage.delay_use=0 reassign_prefmem varlog_size=4096 net.ifnames=0 log_buf_len=1M cmdline-aboot-end tsc=reliable pcie_ports=native rhash_entries=1 usb-storage.delay_use=0 reassign_prefmem iommu=on intel_iommu=on intel_idle.max_cstate=0 varlog_size=4096 sonic.mode=fixed log_buf_len=1M security=apparmor apparmor=1 rw net.ifnames=0 systemd.unified_cgroup_hierarchy=0 quiet systemd.show_status=auto hwaddr_ma1=98:5d:82:1b:0a:2c root=UUID=2f78dbb3-fbea-40f9-b14d-a47a8ecba076 loop=image-20201231.68/fs.squashfs loopfstype=squashfs
$ 
$ cat temp_cmdline | sed -E 's/^(.* sid=[^ ]+).*$/\1/' | tr ' ' '\n'
reboot=p
console=ttyS0
acpi=on
Aboot=Aboot-norcal7-7.2.0-pcie2x4-6128821
block_flash=pci0000:00/0000:00:1f.2/.*host./target0:0:0/.*$
block_drive=pci0000:00/0000:00:1f.2/.*host./target2:0:0/.*$
net_ma2=pci0000:00/0000:00:01.0/.*$
net_ma1=pci0000:00/0000:00:01.1/.*$
block_usb2=pci0000:00/0000:00:14.0/\(usb3/3-2\|usb4/4-2\)/.*$
block_usb1=pci0000:00/0000:00:14.0/\(usb3/3-3\|usb4/4-5\)/.*$
platform=rook
scd.lpc_irq=7
scd.lpc_res_addr=0xb0000000
scd.lpc_res_size=0x10000
sid=Gardena
$ 
```

**Review by vaibhavhd — COMMENTED** [review]
Added some comments. I also  want to make sure this addresses the path where device is running 202012, and has cmdline with duplicate options. Any new installation on this device will remove the duplicate options.
Basically, if we boot from a buggy image to image with fix. Image with fix should be able to fix bugs sent (duplicate options in params_base or cmdline) from previous buggy image.

File: files/Aboot/boot0.j2
Comment (Staphylo) [reply files/Aboot/boot0.j2:89]: I can remove it, it's not related to this PR but to something else.

File: files/Aboot/boot0.j2
Comment (Staphylo) [reply files/Aboot/boot0.j2:399]: It's not related to this PR, I will remove.

File: files/Aboot/boot0.j2
Comment (Staphylo) [reply files/Aboot/boot0.j2:663]: You are making a good point since I just recalled that log_buf_len was added recently, I should use `varlog_size` instead that has been here since the beginning.
It shouldn't lead to false result since these parameters are never set by Aboot, and should therefore never be before `cmdline-aboot-end`

File: files/Aboot/boot0.j2
Comment (Staphylo) [reply files/Aboot/boot0.j2:674]: Will fix, this was written before I addressed the dirty case.

File: files/Aboot/boot0.j2
Comment (Staphylo) [reply files/Aboot/boot0.j2:676]: The whole point of the `write_default_config()` function is to generate `/host/kernel-params-base` and the initial content of `/host/image-XXX/kernel-cmdline`.
It is therefore expected to only output Aboot specific parameter and discard the others.

**Comment by Staphylo** [comment]
@vaibhavhd thank you for your quick review.
Yes it handles the case that you described and that I called `dirty case` in the `How to verify` section of my PR message.
I tested both cases `201811/201911 -> 202012` and `201811/201911 -> 202012 (bug) -> 202012 (fix)`.

File: files/Aboot/boot0.j2
Comment (vaibhavhd) [reply files/Aboot/boot0.j2:676]: About discarding the params - is there any risk in doing so? What about the params such as  `SONIC_BOOT_TYPE`, which are appended to cmdline by (warm/fast/cold)-boot.

Comment (Staphylo) [reply files/Aboot/boot0.j2:676]: Not at all, if you look at the function `write_regular_configs()` which is the one used when booting in non-secureboot mode it does the following:
```
 -> write_common_configs()
     -> write_default_cmdline()
         this one write the Aboot provided parameter only and is the one your comment is on
     -> write_platform_specific_cmdline()
         this write the cmdline parameters required for the current product for the current image
         each image is therefore free to add/remove platform specific parameter (e.g kernel upgrades requiring new params)
     -> write_image_specific_cmdline()
         this write the cmdline parameters for the current image which are product independent (rootfs location, systemd configs, ...)
 -> write_cmdline()
     this functions writes /host/kernel-params-base and /host/image-XXX/kernel-cmdline
```
It is therefore expected that `write_default_cmdline()` only outputs the Aboot specific parameters.
Everything is re-computed at every reboot beside `warm-reboot` and `fast-reboot`
In the `fast-reboot` script with have 2 paths one for secureboot and one for non-secureboot.
In this case it grabs `/host/image-XXX/kernel-cmdline` (generated by sonic-installer install) and append SONIC_BOOT_TYPE
See https://github.com/Azure/sonic-utilities/blob/05c79ef0842deb93e05e3947c4179cb6e0c486e6/scripts/fast-reboot#L334

**Comment by Staphylo** [comment]
I do not see any issue in porting this to other branches like 202111 or 202205.

However it is only useful if the product started in production with 201811/201911 and is upgrading to a newer version of SONiC skipping 202012.
Products that started with 202012 or newer will not experience the issue.

I therefore do not believe it's going to be the case and would therefore prefer porting this fix to newer releases unless you believe it's necessary.
Please advise.

**Review by vaibhavhd — APPROVED** [review]
LGTM, thanks!

**Comment by vaibhavhd** [comment]
> I do not see any issue in porting this to other branches like 202111 or 202205.
> 
> However it is only useful if the product started in production with 201811/201911 and is upgrading to a newer version of SONiC skipping 202012. Products that started with 202012 or newer will not experience the issue.
> 
> I therefore do not believe it's going to be the case and would therefore prefer porting this fix to newer releases unless you believe it's necessary. Please advise.

201811 to future branch jumps will not be made without first getting to 202012. However, I am not sure about jumps from 201911, in which case we may skip 202012. There isn't enough clarity about that at this point. If there are no concerns, I think replicating fix to other branches will be safest approach.

**Comment by Staphylo** [comment]
> > I do not see any issue in porting this to other branches like 202111 or 202205.
> > However it is only useful if the product started in production with 201811/201911 and is upgrading to a newer version of SONiC skipping 202012. Products that started with 202012 or newer will not experience the issue.
> > I therefore do not believe it's going to be the case and would therefore prefer porting this fix to newer releases unless you believe it's necessary. Please advise.
> 
> 201811 to future branch jumps will not be made without first getting to 202012. However, I am not sure about jumps from 201911, in which case we may skip 202012. There isn't enough clarity about that at this point. If there are no concerns, I think replicating fix to other branches will be safest approach.

Sure, I will work on applying this fix to other branches

**Comment by vaibhavhd** [comment]
@Staphylo, please merge this on 202012, if there aren't any more changes to be added here.

**Comment by Staphylo** [comment]
> @Staphylo, please merge this on 202012, if there aren't any more changes to be added here.

@vaibhavhd I do not have merge permissions so I can't merge this PR. Maybe @sujinmkang or @lguohan?

**Comment by vaibhavhd** [comment]
> > @Staphylo, please merge this on 202012, if there aren't any more changes to be added here.
> 
> @vaibhavhd I do not have merge permissions so I can't merge this PR. Maybe @sujinmkang or @lguohan?

Ah. Just merged the PR, thanks!

**Pull request merged**
